added multisite-workflow-example.qmd (#3541)
PecanProject/pecan: documentation/tutorials/multisite-workflow/multisite-workflow-example.qmd

site_info <- tibble::tribble(
	~id, ~lat, ~lon,  ~elev, ~site_group,
	101, 40.1, -88.2, 10,    "NEON",
	102, 41.2, -87.9, 1980,  "NEON",
	201, 39.9, -90.0, 925,   "Ameriflux",
	202, 38.5, -89.5, 1108,  "Ameriflux"
)
# In a real workflow:
# site_info <- read.csv("site_info.csv")
```

## Load Required Libraries

```{r}
library(dplyr)
library(PEcAn.settings)
```

> **NOTE:** You must provide a 'settings' object (as an R list) before running the code below.
> Or load from file: `settings <- read.settings("settings.xml")`

###(sample settings object for running the notebook for multisitesettings)

```{r}
settings <- list(
  outdir = "pecan_runs",
  run = list(
    start.date = "2002-01-01",
    end.date = "2002-12-31",
    site = list(),
    inputs = list(
      met = list(
        id = 99000000001
      )
    )
  ),
  model = list(
    id = 1000000022,
    type = "SIPNET"
  ),
  pfts = list(
  )
)
settings <- as.Settings(settings)

```

## Filter by one group

```{r}
# Filter for NEON sites (change "NEON" to your desired group)
selected_sites <- site_info %>% filter(site_group == "NEON")  
print(selected_sites)           ## prints demo output (not required in main code) 

> settings <- createMultiSiteSettings(settings, selected_sites) ## populates settings with metadata for each site in the sitegroup

Reviewer: One last tweak: Renaming to keep the template settings object around for use in the next example. Sorry I missed this before -- @harshagr70 I'll apply this right away and merge.

```suggestion
settings_neon <- createMultiSiteSettings(settings, selected_sites) ## populates settings with metadata for each site in the sitegroup
listToXml(settings_neon, tag = "pecan")
```

```

## Group by multiple characterstics

Say you're developing a new parameterization for the conifer PFT and you want to test it specifically at high elevation. By filtering and reassigning, this can be done without any edits to the saved `site_info.csv`:

> **NOTE:** Before running the code below, clear your R workspace and recreate the 'settings' object.
> If 'settings' is already a MultiSettings object from a previous run, this code will fail.
> Always start with a fresh 'settings' (as a single Settings object) to avoid errors.

```{r}

settings$pfts <- settings$pfts |>
  append(list(
    name="new_conifer",
    posterior.files = "/path/to/new_conifer/post.distns.Rdata"))
neon_hi_elev <- site_info |>
  filter(
    site_group == "NEON",
    elev > 1000) |>
  mutate(site.pft = "new_conifer")
settings_neon <- createMultiSiteSettings(settings, neon_hi_elev)

listToXml(settings_neon, tag = "pecan")
```

## Shows settings object after "createMultiSiteSettings(settings, selected_sites)"

```{r}

listToXml(settings, tag = "pecan")

> 

Reviewer: ~~~suggestion
## Shows settings object before and after "createMultiSiteSettings(settings, selected_sites)"

```{r}
listToXml(settings, tag = "pecan")
listToXml(settings_neon, tag = "pecan")
```
~~~

---
title: "PEcAn Multisite Flat-File Workflow Example"
format: html
editor: visual
---

## Introduction

This notebook demonstrates how to set up a multisite workflow in PEcAn that uses a subset of the sites defined in a flat-file (`site_info.csv`), without querying the database.  
You will learn how to read site metadata, filter by group, and create a MultiSettings object for your runs.

Sites stored in the BETY database can be queried by `sitegroup` to retrieve predefined sets of sites. Inside the database a sitegroup is nothing but a name given to a list of siteIDs, so in a flat file context we can do the same thing by adding one or more grouping columns.

## Example `site_info.csv`
Suppose you have a CSV file like this: 

```{r}
# For this demo:
site_info <- tibble::tribble(
	~id, ~lat, ~lon,  ~elev, ~site_group,
	101, 40.1, -88.2, 10,    "NEON",
	102, 41.2, -87.9, 1980,  "NEON",
	201, 39.9, -90.0, 925,   "Ameriflux",
	202, 38.5, -89.5, 1108,  "Ameriflux"
)
# In a real workflow:

> # site_info <- read.csv("site_info.csv")

Reviewer 2: it will be very helpful if you can provide a csv file that you have prepared already or may be a link that is available in pecan repo that a user can use.

```

## Load Required Libraries

```{r}
library(dplyr)
library(PEcAn.settings)
```

> **NOTE:** You must provide a 'settings' object (as an R list) before running the code below.

> > Or load from file: settings <- xmlToList(xmlParse("settings.xml"))

Reviewer: ```suggestion
> Or load from file: `settings <- read.settings("settings.xml")`
```

###(sample settings object for running the notebook for multisitesettings)

```{r}
settings <- list(
  outdir = "pecan_runs",
  database = list(
    bety = list(
      user = "bety",
      password = "bety",

>       host = "localhost",

Reviewer 2: host should be `postgres`.please refer `pecan.xml `generated via web as if you refer `tests` directory there in some of the settings file it is mentioned as `localhost` so it is obvious to make mistake.

Reviewer 2: for your information what database config should be
```
  <database>
    <bety>
      <user>bety</user>
      <password>bety</password>
      <host>postgres</host>
      <port>5432</port>
      <dbname>bety</dbname>
      <driver>PostgreSQL</driver>
    </bety>
  </database>
```

Reviewer: For this example we should be able to remove the `bety` block entirely (and probably the enclosing `database` too, but check me on that)

      dbname = "bety",
      driver = "PostgreSQL"
    )
  ),
  run = list(
    start.date = "2002-01-01",
    end.date = "2002-12-31",
    site = list(),
    inputs = list(
      met = list(
        id = 99000000001
      )
    )
  ),
  model = list(
    id = 1000000022,
    type = "SIPNET"
  ),
  pfts = list(
  )
)

> class(settings) <- "Settings"

Reviewer 2: for this entire block of code why we cann't use a existing pecan.xml ?and then just do `read.settings`

Reviewer: I agree we should use the settings constructors rather than set class directly, but I do like that this approach shows the contents inline rather than reading opaquely from a file. 

For demo purposes it would be nice if the inline example were XML text, but that takes enough parsing steps to be distracting. @AritraDey-Dev what do you think about keeping the list format as it is, but changing this line to:
```suggestion
settings <- as.Settings(settings)
```

---
title: "PEcAn Multisite Flat-File Workflow Example"
format: html
editor: visual
---

## Introduction

This notebook demonstrates how to set up a multisite workflow in PEcAn that uses a subset of the sites defined in a flat-file (`site_info.csv`), without querying the database.  
You will learn how to read site metadata, filter by group, and create a MultiSettings object for your runs.

Sites stored in the BETY database can be queried by `sitegroup` to retrieve predefined sets of sites. Inside the database a sitegroup is nothing but a name given to a list of siteIDs, so in a flat file context we can do the same thing by adding one or more grouping columns.

## Example `site_info.csv`

> ## You can either use this tibble for running demo , or can also use the site_info_example.csv in the document/tutorial/multisite-workflow/site_info_example.csv

Reviewer: This is a thoughtful detail but don't think we want to include it -- the point of defining this table inline is to avoid needing multiple files for the demo, and I'm OK with assuming someone who's ready to run a PEcAn demo is familiar enough with R to be able to figure out how to read a dataframe from a CSV.

Reviewer: ```suggestion
```

Suppose you have a CSV file like this: 

```{r}
# For this demo:
site_info <- tibble::tribble(
	~id, ~lat, ~lon,  ~elev, ~site_group,
	101, 40.1, -88.2, 10,    "NEON",
	102, 41.2, -87.9, 1980,  "NEON",
	201, 39.9, -90.0, 925,   "Ameriflux",
	202, 38.5, -89.5, 1108,  "Ameriflux"
)
# In a real workflow:
# site_info <- read.csv("site_info.csv")
```

## Load Required Libraries

```{r}
library(dplyr)
library(PEcAn.settings)
```

> **NOTE:** You must provide a 'settings' object (as an R list) before running the code below.
> Or load from file: `settings <- read.settings("settings.xml")`

###(sample settings object for running the notebook for multisitesettings)

```{r}
settings <- list(
  outdir = "pecan_runs",

>   database = list(
>     bety = list(
>       user = "bety",
>       password = "bety",
>       host = "postgres",
>       dbname = "bety",
>       driver = "PostgreSQL"
>     )
>   ),

Reviewer: Can skip the db section for simplicity:
```suggestion
```

	202, 38.5, -89.5, 1108,  "Ameriflux"
)
# In a real workflow:
# site_info <- read.csv("site_info.csv")
```

---

Load Required Libraries

```{r}
library(dplyr)
# library(PEcAn.settings) # Uncomment if needed
# library(PEcAn.all)      # Uncomment if needed
```

---

Filter by one group

```{r}
# Filter for NEON sites (change "NEON" to your desired group)
selected_sites <- site_info %>% filter(site_group == "NEON")  
print(selected_sites)           ## prints demo output (not required in main code) 
settings <- createMultiSiteSettings(settings, selected_sites) ## populates settings with metadata for each site in the sitegroup

```

---

## Group by multiple characterstics

Say you're developing a new parameterization for the conifer PFT and you want to test it specifically at high elevation. By filtering and reassigning, this can be done without any edits to the saved `site_info.csv`:

```{r}
settings$pfts <- settings$pfts |>
  append(list(
    name="new_conifer",
    posterior.files = "/path/to/new_conifer/post.distns.Rdata"))
neon_hi_elev <- site_info |>
  filter(
    site_group == "NEON",
    elev > 1000) |>
  mutate(site.pft = "new_conifer")

> settings <- createMultiSiteSettings(settings, neon_hi_elev)

Reviewer: ```suggestion
settings_neon <- createMultiSiteSettings(settings, neon_hi_elev)
```

```

---

Example of how settings will look like before and after the execution of "createMultiSiteSettings()" 

Settings object before the execution of 'createMultiSiteSettings()'

```{r}
# Example: settings object with no site info yet

settings <- list(
  run = list(
    start.date = "2025-01-01",
    end.date = "2025-12-31",
    inputs = list()
    # No site info yet!
  )
)

str(settings)  

> 

Reviewer: This can be simplified further by keeping the settings object we created above and displaying that -- we usually show it in XML form just to emphasize the connection between the settings object in your R session and the file written out.
```suggestion
listToXml(settings, tag = "pecan")
```

Reviewer: This also makes me realize the `tag = "pecan"` is boilerplate and I'll submit a PR to let us skip it.

```

Here this shows the settings object after the execution of "createMultiSiteSettings()"

```{r}
# Example: settings object with site info

settings <- list(
  run = list(
    list(
      site = list(id = 101, lat = 40.1, lon = -88.2, elev = 10, site_group = "NEON", met.start = "2025-01-01", met.end = "2025-12-31"),
      start.date = "2025-01-01",
      end.date = "2025-12-31",
      inputs = list()
    ),
    list(
      site = list(id = 102, lat = 41.2, lon = -87.9, elev = 1980, site_group = "NEON", met.start = "2025-01-01", met.end = "2025-12-31"),
      start.date = "2025-01-01",
      end.date = "2025-12-31",
      inputs = list()
    ),
    list(
      site = list(id = 201, lat = 39.9, lon = -90.0, elev = 925, site_group = "Ameriflux", met.start = "2025-01-01", met.end = "2025-12-31"),
      start.date = "2025-01-01",
      end.date = "2025-12-31",
      inputs = list()
    ),
    list(
      site = list(id = 202, lat = 38.5, lon = -89.5, elev = 1108, site_group = "Ameriflux", met.start = "2025-01-01", met.end = "2025-12-31"),
      start.date = "2025-01-01",
      end.date = "2025-12-31",
      inputs = list()
    )
  )
)

> str(settings)

Reviewer: ```suggestion
listToXml(settings_neon, tag = "pecan")
```

---
title: "PEcAn Multisite Flat-File Workflow Example"
format: html
editor: visual
---

## Introduction

This notebook demonstrates how to use a flat-file (`site_info.csv`) to set up a multisite workflow in PEcAn, without querying the database.  

> You will learn how to read site metadata, filter by group, and create a MultiSettings object for your runs.

Reviewer: ```suggestion
This notebook demonstrates how to set up a multisite workflow in PEcAn that uses a subset of the sites defined in a flat-file (`site_info.csv`), without querying the database.  
You will learn how to read site metadata, filter by group, and create a MultiSettings object for your runs.

Sites stored in the BETY database can be queried by `sitegroup` to retrieve predefined sets of sites. Inside the database a sitegroup is nothing but a name given to a list of siteIDs, so in a flat file context we can do the same thing by adding one or more grouping columns.
```

---

## Example `site_info.csv`

Suppose you have a CSV file like this: 

| id  | lat   | lon    | site_group |
|-----|-------|--------|------------|
| 101 | 40.1  | -88.2  | NEON       |
| 102 | 41.2  | -87.9  | NEON       |
| 201 | 39.9  | -90.0  | Ameriflux  |
| 202 | 38.5  | -89.5  | Ameriflux  |

> Save this as `site_info.csv` in your working directory.

Reviewer: You can turn this into a self-contained example by showing the file as a tribble:
~~~suggestion
```{r}
# For this demo:
site_info <- tibble::tribble(
	~id, ~lat, ~lon,  ~elev, ~site_group,
	101, 40.1, -88.2, 10,    "NEON",
	102, 41.2, -87.9, 1980,  "NEON",
	201, 39.9, -90.0, 925,   "Ameriflux",
	202, 38.5, -89.5, 1108,  "Ameriflux"
)
# In a real workflow:
# site_info <- read.csv("site_info.csv")
```
~~~

---

## Load Required Libraries

```{r}
library(dplyr)
# library(PEcAn.settings) # Uncomment if needed
# library(PEcAn.all)      # Uncomment if needed
```

---

## Add the following lines to your workflow code .

```{r}

> site_info <- read.csv("site_info.csv")

Reviewer: ```suggestion
## Filter by one group

```{r}
```

# Filter for NEON sites (change "NEON" to your desired group)

> selected_sites <- site_info %>% filter(site_group == "NEON")  

Reviewer 2: I also worked on this part to retrieve site info based on the site_group while generating the pecan.xml from Quarto, although I used a direct DB query to do this.

A few doubts about your implementation:
I'm currently retrieving site info based on both the `model` and the `site_group`, which is also how it's done via the web interface. But how do you plan to print all the site info based only on `site_group`, given that the model is also required in that case? Are you assuming a default model in such scenarios?

Author: @AritraDey-Dev . when i looked in the code in the step particularly where  createSitegroupMultiSettings( ) is called i dont see the database being queried for the sites on the basis of models as well , maybe in the further steps this might happen , can you pinpoint the part where this thing happens particularly !!

Reviewer 2: Yes, you're right — the goal was to avoid the DB query.So your approach definitely makes it cleaner to retrieve the site info.
That said, I might be mistaken, but I had the impression that the site info was originally being derived based on both the model and the sitegroup .I need to double check this thing

Reviewer: The filtering by model isn't inherent to the concept of sitegroups -- the web workflow does it to support the "conversion" checkbox, which toggles the map between showing all sites in a group or only those that already have inputs (which are model-specific) prepared. I now run quite a few workflows that start from the list of sites and then use it to prepare the missing inputs, so for those cases I wouldn't want any sites filtered out of the group.

In a DB-free notebook context, filtering by input status would probably look something like `site_info$inputs_ready <- some_maybe_expensive_lookup_of_input_files_for_each_site(site_info); selected_sites <- site_info |> filter(site_group == "NEON" & inputs_ready)`. My first thought is that this notebook doesn't need to go into that level of detail, but am curious what y'all think.

print(selected_sites)           ## prints demo output (not required in main code) 

> settings <- createMultiSiteSettings(settings, selected_sites) ## populates settings with metadata for each site in the sitegroup

Reviewer 2: How will settings object look like after this step?Could you please give an example

Author: Adding these lines will just simply skip the db call for the sites thats it , the format of the settings object will remain same as it is being now handled by the createMultiSiteSettings() function .

Reviewer: I agree it's worth showing a very simple before and after example, even though it'll work the same way it does from a DB query. @harshagr70 let me know if you want help constructing this.

```

> 

Reviewer: ~~~suggestion

## Group by multiple characteristics

Say you're developing a new parameterization for the conifer PFT and you want to test it specifically at high elevation. By filtering and reassigning, this can be done without any edits to the saved `site_info.csv`:

```{r}
settings$pfts <- settings$pfts |>
  append(list(
    name="new_conifer",
    posterior.files = "/path/to/new_conifer/post.distns.Rdata"))
neon_hi_elev <- site_info |>
  filter(
    site_group == "NEON",
    elev > 1000) |>
  mutate(site.pft = "new_conifer")
settings <- createMultiSiteSettings(settings, neon_hi_elev)
```
~~~

---

## (Optional) Integrate with workflow 

```{r}

> settings$site_info <- selected_sites

Reviewer: This isn't quite right -- after CreateMultiSiteSettings the site info is already integrated into the settings object, but inside each site rather than as its own object. For example a downstream function looking for site latitude will look at `settings$run[[siteID]]$site$lat`.

Author: True !! , this one seems to be not correct here , will remove it !
